src/v/cloud_storage/segment_meta_cstore.cc

```diff
@@ -563,65 +743,159 @@ class segment_meta_cstore::impl
+            // _write_buffer.erase would do nothing
+            _write_buffer.erase(std::next(m_it), not_replaced_segment);
+        }
 
-    auto inflated_actual_size() const { return _col.inflated_actual_size(); }
+        if (_write_buffer.size() > max_buffer_entries) {
+            flush_write_buffer();
+        }
+    }
 
-    size_t size() const { return _col.size(); }
+    auto inflated_actual_size() const {
+        // TODO how to deal with write_buffer? is it ok to return an approx
+        // value by not flushing?
+        return _col.inflated_actual_size();
+    }
 
-    bool empty() const { return _col.empty(); }
+    size_t size() const {
+        flush_write_buffer();
+        return _col.size();
+    }
+
+    bool empty() const { return _write_buffer.empty() && _col.empty(); }
 
-    bool contains(model::offset o) { return _col.contains(o()); }
+    bool contains(model::offset o) {
+        return _write_buffer.contains(o) || _col.contains(o());
+    }
 
     void prefix_truncate(model::offset new_start_offset) {
+        auto new_begin = _write_buffer.lower_bound(new_start_offset);
+        _write_buffer.erase(_write_buffer.begin(), new_begin);
         _col.prefix_truncate(new_start_offset());
     }
 
     std::unique_ptr<segment_meta_materializing_iterator::impl>
     at_index(size_t ix) const {
+        flush_write_buffer();
         return std::make_unique<segment_meta_materializing_iterator::impl>(
           _col.at_index(ix));
     }
 
-    auto serde_fields() { return std::tie(_col); }
+    auto serde_fields() {
+        flush_write_buffer();
+        return std::tie(_col);
+    }
 
 private:
-    column_store _col;
+    void flush_write_buffer() const {
+        if (_write_buffer.empty()) {
+            return;
+        }
+        auto last_committed_offset = _col.last_committed_offset().value_or(
+          model::offset::min());
+        // find the item separating "Replace" entries (those with a base offset
+        // lower or equal than last_committed_offset) and "Append" entries (all
+        // the other with a base offset greater than last_committed_offset, that
+        // can be simply appended)
+        auto partition_point = _write_buffer.upper_bound(last_committed_offset);
+
+        auto replace_entries_begin = _write_buffer.begin();
+        auto replace_entries_end = partition_point;
+        _col.replace_entries(replace_entries_begin, replace_entries_end);
+
+        auto append_entries_begin = partition_point;
+        auto append_entries_end = _write_buffer.end();
+        // TODO modify append to operate on a range transactionally
+        std::for_each(append_entries_begin, append_entries_end, [&](auto& e) {
+            auto& [k, v] = e;
+            _col.append(v);
+        });
+
+        _write_buffer.clear();
+    }
+
+    mutable absl::btree_map<model::offset, segment_meta> _write_buffer{};
+    mutable column_store _col{};
```

Comment:
its very rare to see mutable members. why do we need this?

Reply:
It's to be able to maintain the const qualifiers on members. Logically the state of the object does not change, but I need to flush to be able to call begin().
Truth to be told, I added flush to be able to get something working correctly to test; there is a plan to expand the iterator to perform the merging on the fly, so that and some other improvements would enable to remove mutable

Reply:
thanks for the explanation